loki/transform/transform_hoist_variables.py

```diff
@@ -0,0 +1,333 @@
+        real, intent(inout) :: array(a)
+        ...
+    end subroutine kernel
+
+using
+
+.. code-block:: python
+
+    # Transformation: Analysis
+    scheduler.process(transformation=HoistTemporaryArraysAnalysis(), reverse=True)
+    # Transformation: Synthesis
+    scheduler.process(transformation=HoistVariablesTransformation())
+
+
+To achieve this two transformation are necessary, whereas the first one is responsible for the *Analysis* and the
+second one for the *Synthesis*. Two base classes
+
+* :class:`.HoistVariablesAnalysis` - *Analysis* part, to be processed in reverse
+    * specialise/implement :func:`find_variables<HoistVariablesAnalysis.find_variables>`
+* :class:`.HoistVariablesTransformation`- *Synthesis* part
+    * specialise/implement :func:`driver_variable_declaration<HoistVariablesSynthesis.driver_variable_declaration>`
+
+are provided to create derived classes for specialisation of the actual hoisting.
+
+.. warning::
+    :class:`.HoistVariablesAnalysis` ensures that all local variables are hoisted!
+    Please consider using a specialised class like :class:`.HoistTemporaryArraysAnalysis` or create a derived class
+    yourself.
+
+.. note::
+    If several of these transformations are carried out in succession, provide a unique ``key`` for each corresponding
+    *Analysis* and *Synthesis* step!
+
+    .. code-block:: python
+
+        key = "UniqueKey"
+        scheduler.process(transformation=HoistTemporaryArraysAnalysis(dim_vars=('b',), key=key), reverse=True)
+        scheduler.process(transformation=HoistTemporaryArraysTransformation(key=key))
+        key = "AnotherUniqueKey"
+        scheduler.process(transformation=HoistTemporaryArraysAnalysis(dim_vars=('a',), key=key), reverse=True)
+        scheduler.process(transformation=HoistTemporaryArraysTransformationAllocatable(key=key))
+"""
+from loki import ir, as_tuple, FindVariables, is_iterable
+from loki.visitors import Transformer, FindNodes
+from loki.transform.transformation import Transformation
+import loki.expression.symbols as sym
+
+
+__all__ = ['HoistVariablesAnalysis', 'HoistVariablesTransformation',
+           'HoistTemporaryArraysAnalysis', 'HoistTemporaryArraysTransformationAllocatable']
+
+
+class HoistVariablesAnalysis(Transformation):
+    """
+    **Base class** for the *Analysis* part of the hoist variables functionality/transformation.
+
+    Traverses all subroutines to find the variables to be hoisted.
+    Create a derived class and override :func:`find_variables<HoistVariablesAnalysis.find_variables>`
+    to define which variables to be hoisted.
+
+    .. note::
+        To be applied **reversed**, in order to recursively find all variables to be hoisted.
+
+    Parameters
+    ----------
+    key : str
+        Access identifier/key for the ``item.user_data`` dictionary. Only necessary to provide if several of
+        these transformations are carried out in succession.
+    """
+
+    _key = 'HoistVariablesTransformation'
```

Comment:
Problem with:

```py
class HoistTemporaryArraysAnalysis(Transformation):
    _key = HoistTemporaryArraysAnalysis.__class__.__name__
    
```

as `NameError: name 'HoistTemporaryArraysAnalysis' is not defined`.

Further problems with derived classes using the `__init__` method and different behaviour for interactive sessions and non-interactive sessions. Therefore proposed solution as implemented.